Source/AwsCommonRuntimeKit/mqtt/Mqtt5.swift

```diff
@@ -0,0 +1,1053 @@
+    let serverKeepAliveSec: UInt16
+
+    /// Whether the server supports retained messages.
+    let retainAvailable: Bool
+
+    /// Whether the server supports wildcard subscriptions.
+    let wildcardSubscriptionsAvailable: Bool
+
+    /// Whether the server supports subscription identifiers
+    let subscriptionIdentifiersAvailable: Bool
+
+    /// Whether the server supports shared subscriptions
+    let sharedSubscriptionsAvailable: Bool
+
+    /// Whether the client has rejoined an existing session.
+    let rejoinedSession: Bool
+
+    /// The final client id in use by the newly-established connection.  This will be the configured client id if one was given in the configuration, otherwise, if no client id was specified, this will be the client id assigned by the server.  Reconnection attempts will always use the auto-assigned client id, allowing for auto-assigned session resumption.
+    let clientId: String
+
+    init (maximumQos: QoS, sessionExpiryIntervalSec: UInt32, receiveMaximumFromServer: UInt16, maximumPacketSizeToServer: UInt32,
+        topicAliasMaximumToServer: UInt16, topicAliasMaximumToClient: UInt16, serverKeepAliveSec: UInt16, retainAvailable: Bool,
+        wildcardSubscriptionsAvailable: Bool, subscriptionIdentifiersAvailable: Bool, sharedSubscriptionsAvailable: Bool, rejoinedSession: Bool,
+        clientId: String) {
+            self.maximumQos = maximumQos
+            self.sessionExpiryIntervalSec = sessionExpiryIntervalSec
+            self.receiveMaximumFromServer = receiveMaximumFromServer
+            self.maximumPacketSizeToServer = maximumPacketSizeToServer
+            self.topicAliasMaximumToServer = topicAliasMaximumToServer
+            self.topicAliasMaximumToClient = topicAliasMaximumToClient
+            self.serverKeepAliveSec = serverKeepAliveSec
+            self.retainAvailable = retainAvailable
+            self.wildcardSubscriptionsAvailable = wildcardSubscriptionsAvailable
+            self.subscriptionIdentifiersAvailable = subscriptionIdentifiersAvailable
+            self.sharedSubscriptionsAvailable = sharedSubscriptionsAvailable
+            self.rejoinedSession = rejoinedSession
+            self.clientId = clientId
+        }
+}
+
+/// Mqtt5 User Property
+public class UserProperty {
+
+    /// Property name
+    let name: String
+
+    /// Property value
+    let value: String
+
+    init (name: String, value: String) {
+        self.name = name
+        self.value = value
+    }
+}
+
+/// Data model of an `MQTT5 PUBLISH <https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901100>`_ packet
+public class PublishPacket {
+
+    /// The payload of the publish message in a byte buffer format
+    var payload: Data?
```

Comment:
I checked on implementation on what is okay to have as empty/default vs. nil and the optionals are that way because they are truly either there or not there on a received packet. Customers should be using a nil check before attempting to access them because depending on whether it exists or not can make a difference in how they handle those packets. 

In the direction of customer created packets/options/etc, they can set them but would not need to access them afterwards and we will check and apply defaults after they are bound to native and processed by the client there. We should not be applying native level defaults to swift level properties.